### ISSUE TITLE
Restoring backlight state on resume from suspend

### ISSUE THREAD

**Issue by icmor** [pobrn/hid-ite8291r3#2]
Sorry if the question is obvious, but is there a built-in way to do this? I find the backlight brightness is kept between reboots but after suspend the backlight is always at its highest setting.

**Comment by pobrn** [comment]
I will take a look. The backlight is restored on boot by `systemd-backlight` - if you have that.

**Comment by pobrn** [comment]
Can you try the current master branch now?

**Comment by icmor** [comment]
It works but with two caveats:
1. It restores the backlight state from the last shutdown, not from before suspend.
2. There's a five second period when resuming the laptop where the backlight will be at full brightness, before restoring state.
Thanks for responding so quickly, maybe this issue and related commits on another driver will help with the quirks:
https://github.com/tuxedocomputers/tuxedo-keyboard/issues/9
https://github.com/tuxedocomputers/tuxedo-keyboard/pull/42

**Comment by pobrn** [comment]
> 1. It restores the backlight state from the last shutdown, not from before suspend.

I cannot reproduce this behaviour. The correct brightness was restored every time I tested it.

> 2. There's a five second period when resuming the laptop where the backlight will be at full brightness, before restoring state.

I can also see some delay, however, in the kernel message buffer I can see that the brightness is restored less than one second after after the kernel comes to life.

You can run `echo 'file hid-ite8291r3.c +p' | sudo tee /sys/kernel/debug/dynamic_debug/control` to enable more logging, every sent or received byte will be logged. E.g.:
```
17:25:14.498736 kernel: ACPI: PM: Low-level resume complete
17:25:14.498852 kernel: ACPI: EC: EC started
17:25:14.498959 kernel: ACPI: PM: Restoring platform NVS memory
17:25:14.499049 kernel: Enabling non-boot CPUs ...
[...]
17:25:14.508361 kernel: hid_ite8291r3 0003:048D:CE00.0002: usb interface gotten
17:25:14.508836 kernel: hid_ite8291r3 0003:048D:CE00.0002: sent: err = 9, buf = 00 09 02 00 00 00 00 00 00
```

**Comment by icmor** [comment]
Tried disabling tlp and my lockscreen but when resuming the state still reverts to that of the previous boot. Eg. Boot the computer, backlight brightness=0 (from previous session) and color is set to "ffffff" with udev (tried without the rule and just setting it manually through sysfs but no difference). Set backlight brightness to maximum with media keys, then run systemctl suspend. On resume backlight goes full brightness for a couple seconds and then goes back to brightness=0. This happens with any value of brightness from the previous shutdown.

Logs from suspend following your advice: 

```
$ sudo dmesg | grep hid_ite
[  286.757533] hid_ite8291r3 0003:048D:CE00.0002: usb interface gotten
[  286.758083] hid_ite8291r3 0003:048D:CE00.0002: sent: err = 9, buf = 00 09 02 00 00 00 00 00 00
[  291.740279] hid_ite8291r3 0003:048D:CE00.0002: sent: err = 9, buf = 00 09 02 00 00 00 00 00 00
[  296.790186] hid_ite8291r3 0003:048D:CE00.0002: usb interface put
```

Don't know if this is of note, but for those few seconds on resume where the backlight attribute is set to maximum, the color attribute is kept. So if I change it to red, on resume it never reverts to its original value.

**Comment by pobrn** [comment]
What happens if you run `cat /sys/class/leds/*kbd_backlight/brightness` before suspending?

**Comment by icmor** [comment]
I don't know what black magic is at play, but `cat /sys/class/leds/*kbd_backlight/brightness` sets a new brightness to restore. I think the problem is with how the media keys update brightness. If I set the brightness any other way, by modifying the sysfs file with tee or with the brightnessctl utility, then brightness is restored on resume. Say I set the brightness to 50 with the media keys, then I cat /sys/.../brightness, then I set the brightness to 0 with the media keys, suspend and resume, then the brightness will be 50. But if I set it any other way, or if I cat the file again after setting the brightness to 0, then brightness will be 0 on resume.

**Comment by icmor** [comment]
It checks out, if I access the brightness file before suspending, it just works.

```
[Unit]
Description=User suspend actions
Before=sleep.target

[Service]
User=%I
Type=forking
Environment=DISPLAY=:0
ExecStartPre=/usr/bin/cat /sys/class/leds/usb-1-6-2-1::kbd_backlight/brightness
ExecStart=/usr/bin/i3lock

[Install]
WantedBy=sleep.target
```

**Comment by pobrn** [comment]
> I think the problem is with how the media keys update brightness.

Yes, you've hit the nail on the head, that is indeed the problem. The reason why it worked for me, is that I have an ugly hack in the [`qc71_laptop`](https://github.com/pobrn/qc71_laptop) module that updates the keyboard LED device brightness when the function keys are pressed.

**Comment by icmor** [comment]
I have no idea how difficult it would be to fix the few seconds that the backlight turns on before restoring brightness, but on the subject of ugly hacks, I present to you /usr/lib/systemd/system-sleep/backlight-quirks

```
#!/bin/sh

case $1 in
    pre)
	cat /sys/class/leds/usb-1-6-2-1::kbd_backlight/brightness
	cat /sys/class/leds/usb-1-6-2-1::kbd_backlight/color > "$(mktemp /tmp/backlight_color.XXX)"
	echo "000000" > /sys/class/leds/usb-1-6-2-1::kbd_backlight/color
	;;
    post)
	echo "$(cat /tmp/backlight_color.*)" > /sys/class/leds/usb-1-6-2-1::kbd_backlight/color
	rm /tmp/backlight_color.*
	;;
esac
```
